src/utils.rs

```diff
@@ -120,3 +129,31 @@ pub fn hex_string_to_h256(mut s: &str) -> Result<H256, String> {
 		},
 	}
 }
+
+pub fn generate_multiproof_grid_dims(grid: Dimensions, target: Dimensions) -> Option<Dimensions> {
+	let cols = min(grid.cols(), target.cols());
+	let rows = min(grid.rows(), target.rows());
+	if grid.cols().get() % cols != 0 || grid.rows().get() % rows != 0 {
+		return None;
+	}
+
+	Dimensions::new(rows, cols)
+}
+pub(crate) fn extract_kate(extension: &HeaderExtension) -> Option<(u16, u16, H256, Vec<u8>)> {
```

Comment:
`extract_kate`, `OptionalExtension`, and `impl OptionalExtension for HeaderExtension` are a bit unclear to me. Headers are not optional so calling something `OptionalExtension` does not make sense.